src/storage/src/source/kafka.rs

```diff
@@ -729,40 +729,32 @@ impl SourceRender for KafkaSourceConnection {
                 assert!(reader.partition_consumers.is_empty());
                 reader.partition_consumers = consumers;
 
-                let positions = reader.consumer.position().unwrap();
-                let topic_positions = positions.elements_for_topic(&reader.topic_name);
                 let mut snapshot_staged = 0;
 
-                for position in topic_positions {
-                    // The offset begins in the `Offset::Invalid` state in which case we simply
-                    // skip this partition.
-                    if let Offset::Offset(offset) = position.offset() {
-                        let pid = position.partition();
-                        let upper_offset = MzOffset::from(u64::try_from(offset).unwrap());
-                        let upper =
-                            Partitioned::new_singleton(RangeBound::exact(pid), upper_offset);
-
-                        let part_cap = reader.partition_capabilities.get_mut(&pid).unwrap();
-                        part_cap.data.downgrade(&upper);
-
-                        if is_snapshotting {
-                            // The `.position()` of the consumer represents what offset we have
-                            // read up to.
-                            snapshot_staged += offset.try_into().unwrap_or(0u64);
-                            // This will always be `Some` at this point.
-                            if let Some(snapshot_total) = snapshot_total {
-                                // We will eventually read past the snapshot total, so we need
-                                // to bound it here.
-                                snapshot_staged = std::cmp::min(snapshot_staged, snapshot_total);
-                            }
+                for (pid, last_offset) in &reader.last_offsets {
```

Comment:
This diff is easier to read with whitespace suppressed. @petrosagg any subtlety here that I'm missing?

Reply:
Had to ditch this approach. Predictably, there was a subtlety here: transaction control messages. We need to look at librdkafka's position to determine whether we've skipped over some offsets that represent transaction control messages.